instrumentation/apache-httpasyncclient-4.1/src/test/java/io/opentelemetry/instrumentation/hypertrace/apachehttpasyncclient/ApacheAsyncClientInstrumentationModuleTest.java

```diff
@@ -16,138 +16,88 @@
 import org.apache.http.client.methods.HttpGet;
 import org.apache.http.client.methods.HttpPost;
 import org.apache.http.concurrent.FutureCallback;
-import org.apache.http.entity.ContentType;
 import org.apache.http.entity.StringEntity;
 import org.apache.http.impl.nio.client.CloseableHttpAsyncClient;
 import org.apache.http.impl.nio.client.HttpAsyncClients;
-import org.apache.http.message.BasicHeader;
-import org.hypertrace.agent.core.instrumentation.HypertraceSemanticAttributes;
-import org.hypertrace.agent.testing.AbstractInstrumenterTest;
-import org.hypertrace.agent.testing.TestHttpServer;
-import org.hypertrace.agent.testing.TestHttpServer.GetJsonHandler;
+import org.hypertrace.agent.testing.AbstractHttpClientTest;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.BeforeAll;
-import org.junit.jupiter.api.Test;
 
-class ApacheAsyncClientInstrumentationModuleTest extends AbstractInstrumenterTest {
-
-  private static final String JSON = "{\"id\":1,\"name\":\"John\"}";
-  private static final TestHttpServer testHttpServer = new TestHttpServer();
+class ApacheAsyncClientInstrumentationModuleTest extends AbstractHttpClientTest {
 
   private static final CloseableHttpAsyncClient client = HttpAsyncClients.createDefault();
 
+  public ApacheAsyncClientInstrumentationModuleTest() {
+    super(true);
+  }
+
   @BeforeAll
-  public static void startServer() throws Exception {
-    testHttpServer.start();
+  public static void startClient() throws Exception {
     client.start();
   }
 
   @AfterAll
-  public static void closeServer() throws Exception {
-    testHttpServer.close();
+  public static void closeClient() throws IOException {
+    client.close();
   }
 
-  @Test
-  public void getJson()
-      throws ExecutionException, InterruptedException, TimeoutException, IOException {
-    HttpGet getRequest =
-        new HttpGet(String.format("http://localhost:%s/get_json", testHttpServer.port()));
-    getRequest.addHeader("foo", "bar");
-    Future<HttpResponse> futureResponse = client.execute(getRequest, new NoopFutureCallback());
-
-    HttpResponse response = futureResponse.get();
-    Assertions.assertEquals(200, response.getStatusLine().getStatusCode());
+  @Override
+  public Pair<Integer, String> doPostRequest(
+      String uri, Map<String, String> headers, String body, String contentType)
+      throws IOException, ExecutionException, InterruptedException {
+    HttpPost request = new HttpPost();
+    for (String key : headers.keySet()) {
+      request.addHeader(key, headers.get(key));
+    }
+    request.setURI(URI.create(uri));
+    StringEntity entity = new StringEntity(body);
+    entity.setContentType(contentType);
+    request.setEntity(entity);
+    request.addHeader("Content-type", contentType);
+    Future<HttpResponse> responseFuture = client.execute(request, new NoopFutureCallback());
+    HttpResponse response = responseFuture.get();
+    Thread.sleep(200);
```

Comment:
Why do we need this wait?

Reply:
The second span was not getting created without this wait. The test was flaky without this wait, and most times the second span with response body was not there.

Reply:
It response body span should be created when body is read, the abstract test can also wait for 2 spans being reported before doing asserts.